File: ProblemSolving/450DSA/JavaScript/src/dynamicprogramming/LongestCommonSubstring.js
/**
 * @author Anirudh Sharma
 * 
 * Given two strings ‘X’ and ‘Y’, find the length of the longest common substring.
 */
const findLCS = (X, Y) => {
    // Special cases
    if (X === undefined || X.length === 0 || Y === undefined || Y.length === 0) {
        return 0;
    }
    // Lengths of the two strings
    const m = X.length;
    const n = Y.length;
    // Lookup table to store the lcs length
    const lookup = Array.from(Array(m + 1), () => Array(n + 1).fill(0));
    // Length of the final lcs
    let lcsLength = 0;
    for (let i = 0; i <= m; i++) {
        for (let j = 0; j <= n; j++) {
            if (i > 0 && j > 0 && X.charAt(i - 1) === Y.charAt(j - 1)) {
                lookup[i][j] = 1 + lookup[i - 1][j - 1];
                lcsLength = Math.max(lcsLength, lookup[i][j]);
            }
        }
    }
    return lcsLength;
};

const main = () => {
    let X = "ABCDGH";
    let Y = "ACDGHR";
    console.log(findLCS(X, Y));

    X = "ABC";
    Y = "ACB";
    console.log(findLCS(X, Y));
};

main();

File: ProblemSolving/450DSA/Python/src/dynamicprogramming/LongestCommonSubstring.py
"""
@author Anirudh Sharma
 
Given two strings ‘X’ and ‘Y’, find the length of the longest common substring.
"""


def findLCS(X, Y):
    # Special cases
    if X is None or len(X) == 0 or Y is None or len(Y) == 0:
        return 0
    # Lengths of the two strings
    m, n = len(X), len(Y)
    # Lookup table to store the lcs length
    lookup = [[0 for y in range(n + 1)] for x in range(m + 1)]
    # Length of the final lcs
    lcsLength = 0
    for i in range(m + 1):
        for j in range(n + 1):
            if i > 0 and j > 0 and X[i - 1] == Y[j - 1]:
                lookup[i][j] = 1 + lookup[i - 1][j - 1]
                lcsLength = max(lcsLength, lookup[i][j])
    return lcsLength


if __name__ == "__main__":
    X = "ABCDGH"
    Y = "ACDGHR"
    print(findLCS(X, Y))

    X = "ABC"
    Y = "ACB"
    print(findLCS(X, Y))
